core/src/main/scala/higherkindness/mu/rpc/srcgen/avro/AvroSrcGenerator.scala

```diff
@@ -152,40 +138,65 @@ final case class AvroSrcGenerator(
+          buildMethodSignature(name, message.getRequest, message.getResponse).map { content =>
+            comment ++ Seq(content, "")
+          }
+      }
+      result.map(_.flatten)
+    }
 
-    val serviceLines =
-      if (requestLines.isEmpty) Seq.empty
-      else
-        Seq(
-          s"@service($serviceParams) trait ${protocol.getName}[F[_]] {",
-          ""
-        ) ++ requestLines :+ "}"
+    val outputCode = requestLines.map { requests =>
+      val serviceLines =
+        if (requests.isEmpty) Seq.empty
+        else {
+          Seq(
+            s"@service(${serviceParams}) trait ${protocol.getName}[F[_]] {",
+            ""
+          ) ++ requests :+ "}"
+        }
+      packageLines ++ importLines ++ messageLines ++ serviceLines
+    }
 
-    outputPath -> (packageLines ++ importLines ++ messageLines ++ serviceLines)
+    outputPath -> outputCode
   }
 
-  private def parseMessage(name: String, request: Schema, response: Schema): String = {
-    val args = request.getFields.asScala
-    if (args.size > 1)
-      throw ParseException("RPC method has more than 1 request parameter")
-    val requestParam = {
-      if (args.isEmpty)
-        s"$DefaultRequestParamName: $EmptyType"
-      else {
-        val arg = args.head
-        if (arg.schema.getType != Schema.Type.RECORD)
-          throw ParseException("RPC method request parameter is not a record type")
-        s"${arg.name}: ${arg.schema.getFullName}"
-      }
-    }
-    val responseParam = {
-      if (response.getType == Schema.Type.NULL) EmptyType
-      else s"${response.getNamespace}.${response.getName}"
+  def validateRequest(request: Schema): ErrorsOr[String] = {
+    val requestArgs = request.getFields.asScala
+    if (requestArgs.size > 1)
+      (s"RPC method ${request.getName} has more than 1 request parameter").invalidNel
+    if (requestArgs.isEmpty) s"$DefaultRequestParamName: $EmptyType".validNel
+    else {
+      val requestArg = requestArgs.head
+      if (requestArg.schema.getType != Schema.Type.RECORD)
+        (s"RPC method request parameter '${requestArg.name}' has non-record return type '${requestArg.schema.getType}'").invalidNel
+      else s"${requestArg.name}: ${requestArg.schema.getFullName}".validNel
     }
-    s"  def $name($requestParam): F[$responseParam]"
   }
 
-  private case class ParseException(msg: String) extends Exception
+  def validateResponse(response: Schema): ErrorsOr[String] = {
+    if (response.getType == Schema.Type.NULL) EmptyType.validNel
+    else {
+      if (response.getType != Schema.Type.RECORD)
+        (s"RPC method response parameter has non-record return type '${response.getType}'").invalidNel
+      else s"${response.getNamespace}.${response.getName}".validNel
+    }
+  }
 
+  def buildMethodSignature(
+      name: String,
+      request: Schema,
+      response: Schema
+  ): ErrorsOr[String] = {
+    validateRequest(request).andThen(requestParam =>
```

Comment:
As discussed offline, this could be done with `mapN`:

```scala
(validateRequest(request), validateResponse).mapN { case (requestParam, responseParam) =>
  s"  def $name($requestParam): F[$responseParam]"
}

Reply:
ah!  I tried `parMapN` and that didn't work, thanks for the insight :)